dali/pipeline/operators/decoder/host_decoder_fixed_crop.cc

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <opencv2/opencv.hpp>
+#include <tuple>
+#include <vector>
+#include <memory>
+#include "dali/error_handling.h"
+#include "dali/image/image_factory.h"
+#include "dali/pipeline/operators/decoder/host_decoder_fixed_crop.h"
+#include "dali/pipeline/operators/common.h"
+
+namespace dali {
+
+HostDecoderFixedCrop::HostDecoderFixedCrop(const OpSpec &spec)
+  : HostDecoder(spec)
+  , CropAttr(spec)
+  , per_sample_crop_window_generators_(batch_size_) {
+  for (int i = 0; i < batch_size_; i++) {
+    DALI_ENFORCE(crop_height_[i] > 0 && crop_width_[i],
+      "crop window dimensions not provided");
+  }
+}
+
+void HostDecoderFixedCrop::SetupSharedSampleParams(SampleWorkspace *ws) {
+  const auto data_idx = ws->data_idx();
+  const auto crop_x_norm = spec_.GetArgument<float>("crop_pos_x", ws, data_idx);
+  const auto crop_y_norm = spec_.GetArgument<float>("crop_pos_y", ws, data_idx);
+
+  per_sample_crop_window_generators_[data_idx] =
+    [this, data_idx, crop_x_norm, crop_y_norm](int H, int W) {
+      CropWindow crop_window;
+      crop_window.h = crop_height_[data_idx];
+      crop_window.w = crop_width_[data_idx];
+      std::tie(crop_window.y, crop_window.x) =
+        CalculateCropYX(
+          crop_y_norm, crop_x_norm,
+          crop_window.h, crop_window.w,
+          H, W);
+      DALI_ENFORCE(crop_window.IsInRange(H, W));
+      return crop_window;
+    };
+}
+
+DALI_REGISTER_OPERATOR(HostDecoderFixedCrop, HostDecoderFixedCrop, CPU);
+
+DALI_SCHEMA(HostDecoderFixedCrop)
```

Comment:
We we cannot have one operator to cover HostDecoderFixedCrop and HostDecoderExternalCrop?
Using AddOptionalArg you can specify that argument is or scalar or tensor. So it can be defined during operator instantiation or be tensor provided from other operator.
So basically HostDecoderExternalCrop logic should cover HostDecoderFixedCrop with minor changes. You can refer to "mask" parameter in DisplacementFilter.

Reply:
As we agreed, Crop/Slice needs unification but for now, the best would be to keep consistency so I will do the following:

- Rename `HostDecoderFixedCrop` to `HostDecoderCrop` as it has the same API as `Crop` (fixed cropping window dimensions and argument input for crop window anchor)
- Rename `HostDecoderExternalCrop` to `HostDecoderSlice` as it has the same API as `Slice` (regular inputs for crop window dimensions and anchor)